sw/device/lib/dif/dif_spi_device.h

```diff
@@ -82,111 +136,180 @@ typedef enum dif_spi_device_result {
 extern const uint16_t kDifSpiDeviceBufferLen;
 
 /**
- * Configuration for initializing a SPI device.
+ * Hardware instantiation parameters for SPI.
+ *
+ * This struct describes information about the underlying hardware that is
+ * not determined until the hardware design is used as part of a top-level
+ * design.
+ */
+typedef struct dif_spi_device_params {
+  /**
+   * The base address for the SPI hardware registers.
+   */
+  mmio_region_t base_addr;
+} dif_spi_device_params_t;
+
+/**
+ * Runtime configuration for SPI.
+ *
+ * This struct describes runtime information for one-time configuration of the
+ * hardware.
  */
 typedef struct dif_spi_device_config {
   dif_spi_device_edge_t clock_polarity;
   dif_spi_device_edge_t data_phase;
   dif_spi_device_bit_order_t tx_order;
   dif_spi_device_bit_order_t rx_order;
   uint8_t rx_fifo_timeout;
+  /**
+   * The length, in bytes, that should be reserved for the RX FIFO.
+   *
+   * `kDifSpiDeviceBufferLen / 2` is a good default for this value.
+   */
   uint16_t rx_fifo_len;
+  /**
+   * The length, in bytes, that should be reserved for the TX FIFO.
+   *
+   * `kDifSpiDeviceBufferLen / 2` is a good default for this value.
+   */
   uint16_t tx_fifo_len;
 } dif_spi_device_config_t;
 
 /**
- * State for a particular SPI device.
+ * A handle to a SPI device.
  *
- * Its member variables should be considered private, and are only provided so
- * that callers can allocate it.
+ * This type should be treated as opaque by users.
  */
 typedef struct dif_spi_device {
-  mmio_region_t base_addr;
-  uint16_t rx_fifo_base;
+  dif_spi_device_params_t params;
   uint16_t rx_fifo_len;
```

Comment:
Shouldn't these be in `dif_spi_device_params_t` instead?

Reply:
They can't, unfortunately. They're not hardware parameters (which is what goes in `params`) but rather runtime parameters (which is what goes in `config`). They need to be kept around as state, because they're used by SW for read/write operations, but are also used by HW for determining how to fill the FIFOs.

Reply:
Thank you for clarification.